src/foam/core/Property.js

```diff
@@ -534,8 +534,10 @@ foam.CLASS({
           // If this is the result of a factory setting the initial value,
           // then don't fire a property change event, since it hasn't
           // really changed.
-          if ( ! factory || oldValue !== undefined )
+          if ( ! factory || oldValue !== undefined ) {
+            if ( prop.final ) prop.visibility = 'RO';
```

Comment:
We have an `if ( isFinal ) { ... }` block a few lines above, wouldn't this make more sense there?

Reply:
I agree. Just updated the code.